### PR TITLE
Tcomm filtering

### PR DESCRIPTION
**Pull request by Consul99**
## About The Pull Request
Adds back tcomm filtering after a 6+ year hiatus. No longer implemented in an overkill embedded scripting language.
Double brackets are used to define a ((target)) to replace with an output which is separated from the trigger by >>. 
/s inserts a separator which can be used if you really need more brackets or > for some reason. An output of only /d will cause the message to be dropped. Filtering supports regex so you'll need to use \ escape for brackets, full stops and other regex operators.

**The pre-filtered messages are always logged on the monitor, so any antag will need to also delete the messages or risk being caught that way.**

Example filter: ".\*clown.\*((killing)).\*>>hugging"
The trigger for the filter is ".\*clown.\*((killing)).\*"
The target for the filter is "killing"
The output for the filter is "hugging"
Since .\* in regex is replaced by any number of any characters, effectively any message where clown and killing is mentioned, killing will be replaced with hugging.
"Help the clown is killing me" -> "Help the clown is hugging me"
Another example: "((/s(This PR is ).\*))>>$1based"
"This PR is not going to be merged anytime this year" -> "This PR is based"

## Why It's Good For The Game
Since most of the game is about communicating with other people, this opens up a whole new way to completely confuse the shit out of everyone while laughing at the consequences.

## Changelog
:cl:
add: tcomms message filtering
/:cl:

### CODE REVIEW

**Comment by mudzbe** [comment]
Cool PR. Which tcomms machine do you use  to do this?

**Comment by Consul99** [comment]
> Cool PR. Which tcomms machine do you use to do this?

Server monitor sets the filters

File: austation/code/game/machinery/telecomms/computers/logbrowser.dm [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+			if("mainmenu")
+				screen = 0
+
+			if("scan")
+				if(servers.len > 0)
+					temp = "<font color = #D70B00>- FAILED: CANNOT PROBE WHEN BUFFER FULL -</font color>"
+
+				else
+					for(var/obj/machinery/telecomms/server/T in urange(25, src))
```
Comment (TheFakeElon) [inline austation/code/game/machinery/telecomms/computers/logbrowser.dm:169]:
```suggestion
					for(var/obj/machinery/telecomms/server/T in GLOB.machines))
						if(get_dist(src, T) > 25)
							continue
```

**Review by TheFakeElon — COMMENTED** [review]
I would strongly prefer it if you could move this to TGUI, web panels are much slower and a bitch to maintain

**Comment by Consul99** [comment]
> I would strongly prefer it if you could move this to TGUI, if you know how. Web panels are much slower, less secure and a bitch to maintain

I don't really know how. I had a look at tgui but it looks way more complicated with the javascript file and all that. Also the UI of the computer which I'm using is already written in html so I'd have to rewrite all that as well. I might get around to it at some point if I learn tgui properly.

File: austation/code/game/machinery/telecomms/computers/logbrowser.dm [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+						race = "Artificial Life"
+
+					else if(isobj(mobtype))
+						race = "Machinery"
+
+					else if(ispath(mobtype, /mob/living/simple_animal))
+						race = "Domestic Animal"
+
+					else
+						race = "<i>Unidentifiable</i>"
```
Comment (TheFakeElon) [inline austation/code/game/machinery/telecomms/computers/logbrowser.dm:75]: Use a switch statement here

**Review by TheFakeElon — COMMENTED** [review]
If you're only overriding a few bits in the proc it might just be cleaner to put it in the non-modular file with comments

**Comment by Consul99** [comment]
> If you're only overriding a few bits in the proc it might just be cleaner to put it in the non-modular file with comments

that's how I originally had it, but I thought it would make more sense to overwrite it incase it ever gets changed, then you don't get conflicts

**Comment by TheFakeElon** [comment]
True! But then the entire proc might break because it's intended function changed, in which case a conflict would actually be better ;)

File: code/game/machinery/telecomms/computers/logbrowser.dm [hunk 3]
```diff
@@ -182,12 +210,23 @@
...
+			if(screen == 2)
+				var/datum/comm_filter_entry/D = SelectedServer.filter_entries[text2num(href_list["delete"])]
+
+				temp = "<font color = #336699>- DELETED FILTER: \[[D.filter_text]\] -</font color>"
+
+				SelectedServer.filter_entries.Remove(D)
+				qdel(D)
+
+			else
+			// austation end -- PR #4379
```
Comment (TheFakeElon) [inline code/game/machinery/telecomms/computers/logbrowser.dm:223]: changes actually end a few more lines down, at qdel.

File: code/game/machinery/telecomms/machines/server.dm [hunk 4]
```diff
@@ -52,6 +52,26 @@
...
+	if(filter_entries)
+		for(var/datum/comm_filter_entry/F in filter_entries)
+			var/message = signal.data["message"]
+			var/regex/R1 = new("[F.trigger]","i")
+			R1.Find(message)
+			if(R1.match)
+				if(F.output == "/d")
+					return
+				F.output = replacetext(F.output,"/d","")
+				var/regex/R2 = new("[F.target]","ig")	// if the target ends with .* the full stop at the end of the message will be eaten
```
Comment (TheFakeElon) [inline code/game/machinery/telecomms/machines/server.dm:65]: regex should use `regex()` not `new()`

**Review by TheFakeElon — APPROVED** [review]

**Pull request merged**